File: trifecta-ios/Trifecta/Features/Thread/ComposerTriggerLogic.swift
import Foundation

enum ComposerTriggerKind: Equatable, Sendable {
    case path
    case slashCommand
    case skill
}

struct ComposerTrigger: Equatable, Sendable {
    var kind: ComposerTriggerKind
    var query: String
    var rangeStart: Int
    var rangeEnd: Int
}

enum ComposerLogic {
    private static func clampCursor(lengthUTF16: Int, _ cursor: Int) -> Int {
        max(0, min(lengthUTF16, cursor))
    }

    private static func isWhitespaceUTF16(_ u: unichar) -> Bool {
        guard let scalar = UnicodeScalar(UInt32(u)) else { return false }
        return CharacterSet.whitespacesAndNewlines.contains(scalar)
    }

    private static func tokenStartUTF16(ns: NSString, cursor: Int) -> Int {
        var index = cursor - 1
        while index >= 0 {
            if isWhitespaceUTF16(ns.character(at: index)) { break }
            index -= 1
        }
        return index + 1
    }

    private static func lineStartUTF16(ns: NSString, cursor: Int) -> Int {
        let capped = clampCursor(lengthUTF16: ns.length, cursor)
        if capped <= 0 { return 0 }
        let search = NSRange(location: 0, length: capped)
        let nl = ns.range(of: "\n", options: .backwards, range: search)
        if nl.location != NSNotFound {
            return nl.location + nl.length
        }
        return 0
    }

    /// UTF-16 offsets, aligned with `UITextView.selectedRange`.
    static func detectTrigger(text: String, cursorUTF16: Int) -> ComposerTrigger? {
        let ns = text as NSString
        let len = ns.length
        let cursor = clampCursor(lengthUTF16: len, cursorUTF16)

        let lineStart = lineStartUTF16(ns: ns, cursor: cursor)
        let lineLen = cursor - lineStart
        if lineLen >= 0, lineStart + lineLen <= len {
            let linePrefix = ns.substring(with: NSRange(location: lineStart, length: lineLen))
            if linePrefix.hasPrefix("/"),
               let regex = try? NSRegularExpression(pattern: #"^\/(\S*)$"#, options: []),
               let match = regex.firstMatch(in: linePrefix, range: NSRange(location: 0, length: (linePrefix as NSString).length)),
               match.numberOfRanges >= 2 {
                let r = match.range(at: 1)
                if r.location != NSNotFound {
                    let commandQuery = (linePrefix as NSString).substring(with: r)
                    return ComposerTrigger(
                        kind: .slashCommand,
                        query: commandQuery,
                        rangeStart: lineStart,
                        rangeEnd: cursor
                    )
                }
            }
        }

        let tokenStart = tokenStartUTF16(ns: ns, cursor: cursor)
        let tokenLen = cursor - tokenStart
        guard tokenLen > 0, tokenStart + tokenLen <= len else { return nil }
        let token = ns.substring(with: NSRange(location: tokenStart, length: tokenLen))

        if token.hasPrefix("$") {
            return ComposerTrigger(
                kind: .skill,
                query: String(token.dropFirst()),
                rangeStart: tokenStart,
                rangeEnd: cursor
            )
        }
        if token.hasPrefix("@") {
            return ComposerTrigger(
                kind: .path,
                query: String(token.dropFirst()),
                rangeStart: tokenStart,
                rangeEnd: cursor
            )
        }
        return nil
    }

    static func replaceRangeUTF16(in text: String, rangeStart: Int, rangeEnd: Int, replacement: String) -> (text: String, cursorUTF16: Int) {
        let ns = text as NSString
        let len = ns.length
        let s = max(0, min(len, rangeStart))
        let e = max(s, min(len, rangeEnd))
        let prefix = ns.substring(to: s)
        let suffix = ns.substring(from: e)
        let next = prefix + replacement + suffix
        let cursor = s + (replacement as NSString).length
        return (next, cursor)
    }

    static func parseStandaloneModeSlash(_ trimmed: String) -> ProviderInteractionMode? {
        let t = trimmed.trimmingCharacters(in: .whitespacesAndNewlines)
        if t.range(of: #"^/plan\s*$"#, options: [.regularExpression, .caseInsensitive]) != nil {
            return .plan
        }
        if t.range(of: #"^/default\s*$"#, options: [.regularExpression, .caseInsensitive]) != nil {
            return .default
        }
        return nil
    }

    static func isStandaloneModelSlash(_ trimmed: String) -> Bool {
        let t = trimmed.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.range(of: #"^/model\s*$"#, options: [.regularExpression, .caseInsensitive]) != nil
    }
}

File: trifecta-android/app/src/main/java/com/belweave/trifecta/features/thread/ComposerTriggerLogic.kt
package com.belweave.trifecta.features.thread

import com.belweave.trifecta.core.models.ProviderInteractionMode

/**
 * Detection result for a composer trigger token (`@`, `$`, `/`).
 *
 * The range is expressed in UTF-16 offsets so it lines up with
 * [androidx.compose.ui.text.input.TextFieldValue.selection] which uses
 * UTF-16 code unit indices internally.
 */
enum class ComposerTriggerKind { PATH, SLASH_COMMAND, SKILL }

data class ComposerTrigger(
    val kind: ComposerTriggerKind,
    val query: String,
    val rangeStart: Int,
    val rangeEnd: Int
)

/** Result returned by [ComposerLogic.replaceRangeUtf16]. */
data class ReplaceResult(val text: String, val cursorUtf16: Int)

object ComposerLogic {

    private fun clampCursor(length: Int, cursor: Int): Int =
        cursor.coerceIn(0, length)

    private fun isWhitespace(c: Char): Boolean = c.isWhitespace()

    private fun tokenStart(text: CharSequence, cursor: Int): Int {
        var index = cursor - 1
        while (index >= 0) {
            if (isWhitespace(text[index])) break
            index -= 1
        }
        return index + 1
    }

    private fun lineStart(text: CharSequence, cursor: Int): Int {
        val capped = clampCursor(text.length, cursor)
        if (capped <= 0) return 0
        val nl = text.lastIndexOf('\n', capped - 1)
        return if (nl >= 0) nl + 1 else 0
    }

    /** UTF-16 offsets, matching `TextFieldValue.selection`. */
    fun detectTrigger(text: String, cursorUtf16: Int): ComposerTrigger? {
        val len = text.length
        val cursor = clampCursor(len, cursorUtf16)

        val ls = lineStart(text, cursor)
        val lineLen = cursor - ls
        if (lineLen >= 0 && ls + lineLen <= len) {
            val linePrefix = text.substring(ls, ls + lineLen)
            if (linePrefix.startsWith("/")) {
                val match = SLASH_LINE_REGEX.matchEntire(linePrefix)
                if (match != null) {
                    val q = match.groupValues.getOrNull(1) ?: ""
                    return ComposerTrigger(
                        kind = ComposerTriggerKind.SLASH_COMMAND,
                        query = q,
                        rangeStart = ls,
                        rangeEnd = cursor
                    )
                }
            }
        }

        val ts = tokenStart(text, cursor)
        val tokenLen = cursor - ts
        if (tokenLen <= 0 || ts + tokenLen > len) return null
        val token = text.substring(ts, ts + tokenLen)
        return when {
            token.startsWith("$") -> ComposerTrigger(
                kind = ComposerTriggerKind.SKILL,
                query = token.drop(1),
                rangeStart = ts,
                rangeEnd = cursor
            )
            token.startsWith("@") -> ComposerTrigger(
                kind = ComposerTriggerKind.PATH,
                query = token.drop(1),
                rangeStart = ts,
                rangeEnd = cursor
            )
            else -> null
        }
    }

    fun replaceRangeUtf16(
        text: String,
        rangeStart: Int,
        rangeEnd: Int,
        replacement: String
    ): ReplaceResult {
        val len = text.length
        val s = rangeStart.coerceIn(0, len)
        val e = rangeEnd.coerceIn(s, len)
        val next = text.substring(0, s) + replacement + text.substring(e)
        val cursor = s + replacement.length
        return ReplaceResult(next, cursor)
    }

    fun parseStandaloneModeSlash(trimmed: String): ProviderInteractionMode? {
        val t = trimmed.trim()
        if (PLAN_REGEX.matches(t)) return ProviderInteractionMode.PLAN
        if (DEFAULT_REGEX.matches(t)) return ProviderInteractionMode.DEFAULT
        return null
    }

    fun isStandaloneModelSlash(trimmed: String): Boolean =
        MODEL_REGEX.matches(trimmed.trim())

    private val SLASH_LINE_REGEX = Regex("""^/(\S*)$""")
    private val PLAN_REGEX = Regex("""^/plan\s*$""", RegexOption.IGNORE_CASE)
    private val DEFAULT_REGEX = Regex("""^/default\s*$""", RegexOption.IGNORE_CASE)
    private val MODEL_REGEX = Regex("""^/model\s*$""", RegexOption.IGNORE_CASE)
}
